src/btree/bt_curnext.c

```diff
@@ -904,8 +904,16 @@ __wt_btcur_next(WT_CURSOR_BTREE *cbt, bool truncating)
         if (page != NULL &&
           (cbt->page_deleted_count > WT_BTREE_DELETE_THRESHOLD ||
             (newpage && cbt->page_deleted_count > 0))) {
-            WT_ERR(__wt_page_dirty_and_evict_soon(session, cbt->ref));
-            WT_STAT_CONN_INCR(session, cache_eviction_force_delete);
+            /* If checkpoint is happening on the btree, we can only evict clean content. */
+            if (__wt_btree_syncing_by_other_session(session)) {
+                if (!__wt_page_is_modified(page)) {
+                    __wt_page_evict_soon(session, cbt->ref);
```

Comment:
Can we write a python test that covers this? I think it may be possible but it would require that the ref->state is read, or another stat increment is tested. Effectively you'd need two threads:

1. The checkpoint thread, which syncs the same b-tree slowly
2. The cursor->next thread that will either dirty the page or not 

The answer to this question may be "we can't test this currently" but then I'd like to know what we need to enable this code becoming testable.

Reply:
I agree that we need to come up with a test to better understand the impact of the changes. I believe we need:

- One table with lots of records per page
- Delete a bunch of them, keep the deleted records in memory (a big cache size should help)
- Trigger a checkpoint (we can make it slow using debug options)
- Start a read scan and we should hit this code path